### PR TITLE
Logon, Logoff & Shutdown Sounds Restored - v1.0.0

### PR DESCRIPTION
**Pull request by TorutheRedFox**
Restores the playback of events once played by Explorer during the start, end, locking and unlocking of a logon session.

### CODE REVIEW

File: mods/logon-logoff-shutdown-sounds.wh.cpp [hunk 1]
```diff
@@ -0,0 +1,582 @@
+// ==WindhawkMod==
+// @id              logon-logoff-shutdown-sounds
+// @name            Logon, Logoff & Shutdown Sounds Restored
+// @description     Restores the logon, logoff and shutdown sounds from earlier versions of Windows
+// @version         1.0.0
+// @author          Toru the Red Fox
+// @github          https://github.com/TorutheRedFox
+// @twitter         https://twitter.com/TorutheRedFox
```
Comment (m417z) [inline mods/logon-logoff-shutdown-sounds.wh.cpp:8]: Please send me (https://x.com/m417z) a direct message with the following content:

> I attest that I'm the sole owner of both this X account (https://x.com/TorutheRedFox) and the following GitHub account: https://github.com/TorutheRedFox

File: mods/logon-logoff-shutdown-sounds.wh.cpp [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+        && szFileName[0])
+    {
+        // flags are relevant, we try to not stomp currently playing sounds
+        PlaySound(szFileName, NULL, SND_FILENAME | SND_ASYNC | SND_NODEFAULT | SND_NOSTOP);
+    }
+}
+
+typedef LRESULT (* _OnSessionChange_t)(void* _this, WPARAM wParam, LPARAM lParam);
+_OnSessionChange_t _OnSessionChange_orig;
+LRESULT _OnSessionChange_hook(void* _this, WPARAM wParam, LPARAM lParam)
```
Comment (m417z) [inline mods/logon-logoff-shutdown-sounds.wh.cpp]: Calling conventions are missing, which is likely to cause a crash for 32-bit targets. Please add calling conventions or, if the mod doesn't support 32-bit, limit it to 64-bit, which will also make compilation faster. See:
https://github.com/ramensoftware/windhawk/wiki/Development-tips#hooks-and-calling-conventions

File: mods/logon-logoff-shutdown-sounds.wh.cpp [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+};
+
+CSoundWnd::CSoundWnd()
+    : m_refCount(1)
+    , m_hwndSound(nullptr)
+    , m_thread(nullptr)
+{
+}
+
+BOOL CSoundWnd::Init()
```
Comment (m417z) [inline mods/logon-logoff-shutdown-sounds.wh.cpp]: I see that some of the code is based on code from ExplorerPatcher. If significant parts are copied or heavily inspired by other projects, consider crediting them and specifying a compatible license for your mod.

File: mods/logon-logoff-shutdown-sounds.wh.cpp [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+
+    LoadSettings();
+
+    g_hExplorer = GetModuleHandleW(NULL);
+    if (!g_hExplorer)
+        Wh_Log(L"Failed to get Explorer's handle");
+
+    g_hShlwapi = LoadLibrary(TEXT("shlwapi.dll"));
+
+    WindhawkUtils::SYMBOL_HOOK aHooks[] = {
```
Comment (m417z) [inline mods/logon-logoff-shutdown-sounds.wh.cpp]: As the message says, rename the hooks variable. In this case, rename it to `explorerExeHooks` or one of the other accepted variants.

**Comment by TorutheRedFox** [comment]
The only thing based on what ExplorerPatcher did iirc is moving the code to wait for the sound to finish to WM_QUERYSESSIONEND, otherwise it's just reverse engineered code from vista's x86 explorer and 7's x64 explorer, and yeah I forgot about x86 I should fix that lol

File: mods/logon-logoff-shutdown-sounds.wh.cpp (hunk 2 above)
Comment (m417z) [reply mods/logon-logoff-shutdown-sounds.wh.cpp]: You only had to add it for hooks, in your case just `_OnSessionChange_hook`, but the additional calling conventions don't harm.

**Pull request merged**